File: archive/c/c-sharp/RomanNumeral.cs
if (args is not [var input])
    return ExitWith("Usage: please provide a string of roman numerals");

if (!TryRomanToInt(input.AsSpan().Trim(), out int value))
    return ExitWith("Error: invalid string of roman numerals");

Console.WriteLine(value);
return 0;

static bool TryRomanToInt(ReadOnlySpan<char> roman, out int result)
{
    result = 0;
    if (roman.Length == 0)
        return true;

    int prev = 0;

    for (int i = roman.Length - 1; i >= 0; i--)
    {
        if (!TryGetValue(roman[i], out int current))
            return false;

        if (current < prev)
            result -= current;
        else
            result += current;

        prev = current;
    }

    return true;
}

static bool TryGetValue(char c, out int value)
{
    value = char.ToUpper(c) switch
    {
        'I' => 1,
        'V' => 5,
        'X' => 10,
        'L' => 50,
        'C' => 100,
        'D' => 500,
        'M' => 1000,
        _ => 0
    };

    return value != 0;
}

static int ExitWith(string message)
{
    Console.WriteLine(message);
    return 1;
}

File: archive/p/pascal/roman_numeral.pas
program RomanNumeral;

{$mode objfpc}{$H+}

uses
   SysUtils;

procedure ShowUsage;
begin
   Writeln('Usage: please provide a string of roman numerals');
   Halt(1);
end;

function RomanToInt(const S: string): int64;
const
   RomanChars: array[0..6] of char = ('I', 'V', 'X', 'L', 'C', 'D', 'M');
   RomanValues: array[0..6] of integer = (1, 5, 10, 50, 100, 500, 1000);
var
   Value: array[char] of integer;
   i, len: integer;
   ch: char;
   Answer, PreviousValue, CurrentValue: int64;
begin
   FillChar(Value, SizeOf(Value), 0);

   for i := Low(RomanChars) to High(RomanChars) do
      Value[RomanChars[i]] := RomanValues[i];

   Answer := 0;
   PreviousValue := 0;
   len := Length(S);

   if len = 0 then
      Exit(0);

   for i := len downto 1 do
   begin
      ch := S[i];
      CurrentValue := Value[ch];
      if CurrentValue = 0 then
         raise Exception.Create('Error: invalid string of roman numerals');

      if CurrentValue < PreviousValue then
         Dec(Answer, CurrentValue)
      else
      begin
         Inc(Answer, CurrentValue);
         PreviousValue := CurrentValue;
      end;
   end;

   Result := Answer;
end;

var
   Input: string;
   Number: int64;
begin
   if ParamCount <> 1 then
      ShowUsage;

   Input := AnsiUpperCase(Trim(ParamStr(1)));

   try
      Number := RomanToInt(Input);
      Writeln(Number);
   except
      on E: Exception do
      begin
         Writeln(E.Message);
         Halt(1);
      end;
   end;
end.
